torch/csrc/autograd/engine.cpp

```diff
@@ -1531,7 +1542,19 @@ auto Engine::ready_queue(
     at::Device device) -> std::shared_ptr<ReadyQueue> {
   bool multithreading_disabled =
       !c10::AutogradState::get_tls_state().get_multithreading_enabled();
-  if (multithreading_disabled || should_run_in_cpu_ready_queue(device.type())) {
+  // Note [ Engine threading optimization when single device ]
+  //
+  // If during graph traversal, we find that there is only a single distinct
+  // device, we will reuse the calling thread. This can result in a perf win
+  // due to fewer cache misses, TLB misses, context switches, etc.
+  //
+  // The only situation current_graph_task is not yet set is when
+  // execute_with_graph_task enqueues the very first task. We manually check
+  // num_distinct_devices_ there.
+  auto only_single_device =
+      current_graph_task && current_graph_task->num_distinct_devices_ == 1;
```

Comment:
current_graph_task could be empty here? :o 
Also taking the graph_task as argument would be the expected behavior rather than delying on a global variable no?